core/drivers/nxp_ls_i2c.c

```diff
@@ -0,0 +1,420 @@
+		ibfd = clock_divisor_pair[clock_dvisor_pair_size - 1].ibfd;
+	} else {
+		for (index = 0; index < clock_dvisor_pair_size; index++) {
+			if (clock_divisor_pair[index].divisor >=
+			    clock_divisor) {
+				ibfd = clock_divisor_pair[index].ibfd;
+				break;
+			}
+		}
+	}
+
+	return ibfd;
+}
+
+TEE_Result i2c_init(struct nxp_i2c_data *i2c_data)
+{
+	struct i2c_regs *regs = NULL;
+	uint16_t clock_divisor = 0;
+	uint8_t ibfd = 0; // I2c Bus Frequency Divider Register
+	size_t size = 0;
+	int node = 0;
+	vaddr_t ctrl_base = 0;
+	void *fdt = NULL;
+
+	/*
+	 * First get the I2C Controller base address from the DTB
+	 * if DTB present and if the I2C Controller defined in it.
+	 */
+	fdt = get_embedded_dt();
+
+	node = fdt_path_offset(fdt,
+			       i2c_controller_map[i2c_data->i2c_controller]);
+	if (node > 0) {
+		if (dt_map_dev(fdt, node, &ctrl_base, &size) < 0) {
+			EMSG("Unable to get virtual address");
+			return TEE_ERROR_GENERIC;
+		}
+	} else {
+		EMSG("Unable to get I2C offset node");
+		return TEE_ERROR_ITEM_NOT_FOUND;
+	}
+
+	i2c_data->base = ctrl_base;
+
+	regs = (struct i2c_regs *)ctrl_base;
+	if (I2C_ERRATUM_A009203_FIX)
+		i2c_erratum_a009203(ctrl_base);
+
+	clock_divisor = (i2c_data->i2c_bus_clock + i2c_data->speed - 1) /
+			i2c_data->speed;
+	ibfd = i2c_get_ibfd(ctrl_base, clock_divisor);
+
+	io_write8((vaddr_t)&regs->ibfd, ibfd);
+
+	i2c_reset(ctrl_base);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result i2c_bus_test_bus_busy(struct i2c_regs *regs, bool test_busy)
+{
+	uint32_t index = 0;
+	uint8_t reg = 0;
+
+	for (index = 0; index < I2C_NUM_RETRIES; index++) {
+		reg = io_read8((vaddr_t)&regs->ibsr);
+
+		if (reg & I2C_IBSR_IBAL) {
+			io_write8((vaddr_t)&regs->ibsr, reg);
+			return TEE_ERROR_BUSY;
+		}
+
+		if (test_busy && (reg & I2C_IBSR_IBB))
+			break;
+
+		if (!test_busy && !(reg & I2C_IBSR_IBB))
+			break;
+
+		mdelay(1);
+	}
+
+	if (index == I2C_NUM_RETRIES)
+		return TEE_ERROR_BUSY;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result i2c_transfer_complete(struct i2c_regs *regs, bool test_rx_ack)
+{
+	uint32_t index = 0;
+	uint8_t reg = 0;
+
+	for (index = 0; index < I2C_NUM_RETRIES; index++) {
+		reg = io_read8((vaddr_t)&regs->ibsr);
+
+		if (reg & I2C_IBSR_IBIF) {
+			// Write 1 to clear the IBIF field
+			io_write8((vaddr_t)&regs->ibsr, reg);
+			break;
+		}
+		mdelay(1);
+	}
+
+	if (index == I2C_NUM_RETRIES)
+		return TEE_ERROR_BUSY;
+
+	if (test_rx_ack && (reg & I2C_IBSR_RXAK))
+		return TEE_ERROR_NO_DATA;
+
+	if (reg & I2C_IBSR_TCF)
+		return TEE_SUCCESS;
+
+	return TEE_ERROR_GENERIC;
+}
+
+static TEE_Result i2c_read(struct i2c_regs *regs, uint32_t slave_address,
+			   struct i2c_operation *operation,
+			   bool is_last_operation)
+{
+	TEE_Result status = TEE_ERROR_GENERIC;
+	uint32_t index = 0;
+
+	// Write Slave Address
+	io_write8((vaddr_t)&regs->ibdr, (slave_address << BIT(0)) | BIT(0));
+	status = i2c_transfer_complete(regs, I2C_BUS_TEST_RX_ACK);
+	if (status != TEE_SUCCESS)
```

Comment:
While it's OK to check like this we normally use the `if (res)` style instead and consistency make it easier to read the code. That way it's easier to focus on what matters.